[DebugInfo][RemoveDIs] Support DPValues in HWAsan

jmorse opened the pull request:
This patch extends HWASAN to support maintenance of debug-info that isn't stored as intrinsics, but is instead in a DPValue object. This is straight-forwards: we collect any such objects in StackInfoBuilder, and apply the same operations to them as we would to dbg.value and similar intrinsics.

I've also replaced some calls to getNextNode with debug-info skipping next calls, and use iterators for instruction insertion rather than instruction pointers. This avoids any difference in output between intrinsic / non-intrinsic debug-info, but also means that any debug-info comes before code inserted by HWAsan, rather than afterwards. See the test modifications, where the variable assignment (presented as a dbg.value) jumps up over all the code inserted by HWAsan. Seeing how the code inserted by HWAsan is always (AFAIUI) given the source-location of the instruction being instrumented, I don't believe this will have any effect on which lines variable assignments become visible on; it may extend the number of instructions covered by the assignments though.

jmorse committed:
[DebugInfo][RemoveDIs] Support DPValues in HWAsan

This patch extends HWASAN to support maintenence of debug-info that isn't
stored as intrinsics, but is instead in a DPValue object. This is
straight-forwards: we collect any such objects in StackInfoBuilder, and
apply the same operations to them as we would to dbg.value and similar
intrinsics.

I've also replaced some calls to getNextNode with debug-info skipping next
calls, and use iterators for instruction insertion rather than instruction
pointers. This avoids any difference in output between intrinsic /
non-intrinsic debug-info, but also means that any debug-info comes before
code inserted by HWAsan, rather than afterwards (see the test
modifications). Seeing how the code inserted by HWAsan is always (AFAIUI)
given the source-location of the instruction being instrumented, I don't
believe this will have any effect on which lines variable assignments
become visible on; it may extend the number of instructions covered by the
assignments though.

fmayer commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -1425,17 +1425,24 @@ bool HWAddressSanitizer::instrumentStack(memtag::StackInfo &SInfo,
+                                                           NewOps, LocNo));
+    };
+
+    for (auto *DDI : Info.DbgVariableIntrinsics)
```
optional nit: 

```
llvm::for_each(Info.DbgVariableIntrinsics, AnnotateDbgRecord);
llvm::for_each(Info.DbgVariableIntrinsics, DbgVariableRecords);
```

fmayer commented on llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:
```diff
@@ -1532,8 +1539,8 @@ void HWAddressSanitizer::sanitizeFunction(Function &F,
 
-  Instruction *InsertPt = &*F.getEntryBlock().begin();
-  IRBuilder<> EntryIRB(InsertPt);
+  BasicBlock::iterator InsertPt = F.getEntryBlock().begin();
```
For my understanding: Can you quickly explain why this (and related changes below) are needed?

jmorse replied:
We're in the process of changing how debug-info variable locations are recorded -- right now they're stored as intrinsics, but in the future we'll be able to attach them directly to instructions, which has a variety of benefits (efficiency and stability) because passes won't need to take account of instructions that aren't "real". As part of that, we need to put a tiny bit of debug-info into the instruction-iterator class, and have most instruction insertions be performed with iterators, hence this particular change on these lines.

I can go into detail, but by using iterators everywhere in scenarios like this, we're able to perform all the maintenance of debug-info in the background, without LLVM passes having to perform it themselves.

(There's a meaningful change that I'll describe in a main comment in a moment,)

fmayer replied:
Thanks for the detailed explanation.

jmorse committed:
Use llvm:for_each for transforming a collection

jmorse committed:
I should probably test things before pushing X_X

jmorse commented:
Just to explain the addition of `getNextNonDebugInstruction` in this patch and the related test changes: in the two flavours of alloca.ll we have:
```
    %x = alloca ty    ; line 1
    call void @llvm.dbg.value
    call some_function ; line 2
```
And that becomes (without the patch):
```
    %x = alloca ty     ; line 1
    hwasan_instrumentation_insts ; line 2
    call void @llvm.dbg.value
    call some_function ; line 2
```

Right now, if someone operating a debugger requests a breakpoint on line 2 then they'll very likely end up stopped on the hwasan instrumentation instructions, and won't be able to see any variable assignments that were supposed to happen before line 2, which creates a small risk that they'll get confused.

Using `getNextNonDebugInstruction` makes the hwasan instrumentation come after the dbg.value intrinsic:
```
    %x = alloca ty     ; line 1
    call void @llvm.dbg.value
    hwasan_instrumentation_insts ; line 2
    call some_function ; line 2
```
Now when someone requests a breakpoint on line 2, they'll still stop on the hwasan instrumentation instructions, but all the variable assignments that happen before line 2 will be visible. It sort of makes the addition of the instrumentation more transparent to the developer.

Normally this sort of change deserves it's own patch -- however it's inherently connected with the debuginfo iterators change, as what we're doing there effectively makes `getNextNode` behave like `getNextNonDebugInstruction` at all times, so I thought it best to do it all at the same time. This'll avoid spurious test differences when we enable the new feature, plus as mentioned, the new output is slightly better.

fmayer commented:
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> You can test this locally with the following command:
> View the diff from clang-format here.

Please fix the formatting

vitalybuka commented:
How will this interact with #78606

OCHyams commented:
> How will this interact with https://github.com/llvm/llvm-project/pull/78606

@jmorse I'm happy for you to land this patch first, and then I will incorporate DPVAssign updates into mine?

jmorse committed:
clang-format

fmayer commented on llvm/lib/Target/AArch64/AArch64StackTagging.cpp:
```diff
@@ -578,6 +578,8 @@ bool AArch64StackTagging::runOnFunction(Function &Fn) {
     // Fixup debug intrinsics to point to the new alloca.
     for (auto *DVI : Info.DbgVariableIntrinsics)
       DVI->replaceVariableLocationOp(OldAI, Info.AI);
+    for (auto *DPV : Info.DbgVariableRecords)
```
This seems untested. I can comment this out and 

```
build/bin/llvm-lit -v llvm/test/CodeGen/AArch64/
```

still passes.

jmorse replied:
It's not tested under the default configuration and so the lack-of-failure is expected. To ensure the code isn't rotting, we've got this [0] buildbot set up to exercise the "new debug-info" configuration, which requires passing `LLVM_EXPERIMENTAL_DEBUGINFO_ITERATORS=On` to cmake. With that enabled, the test RUN lines where `--try-experimental-debuginfo-iterators` has been added will operate under "new" debug-info mode, and  you Should (TM) see a test failure.

We're inches away from turning this on-by-default and it being tested everywhere, after which there won't be this obscure testing arrangement happening any more.

[0] https://lab.llvm.org/buildbot/#/builders/275

fmayer replied:
Oh thanks! I'll keep that in mind. I wanted to change this code a bit so wanted to make sure I'm actually running the tests, now I know how.

fmayer replied:
I re-rean this with `-DLLVM_EXPERIMENTAL_DEBUGINFO_ITERATORS=On` in the CMake but it still passes if I comment out these lines.

jmorse commented:
/me squints -- either we added the coverage to the wrong file, or something else weird has gone wrong. Either way, in 5ce2f73b2e5e I've added another `--try-experimental-debuginfo-iterators` runline to ``CodeGen/AArch64/stack-tagging-dbg.ll, which does in fact fail if you delete the relevant lines. Apologies for the intervening bother.
